### PR TITLE
Review proc macro API 1.2

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rust/issues/38356

Summary of applied changes:
- Documentation for proc macro API 1.2 is expanded.
- Renamed APIs: `Term` -> `Ident`, `TokenTree::Term` -> `TokenTree::Ident`, `Op` -> `Punct`, `TokenTree::Op` -> `TokenTree::Punct`, `Op::op` -> `Punct::as_char`.
- Removed APIs: `Ident::as_str`, use `Display` impl for `Ident` instead.
- New APIs (not stabilized in 1.2): `Ident::new_raw` for creating a raw identifier (I'm not sure `new_x` it's a very idiomatic name though).
- Runtime changes:
    - `Punct::new` now ensures that the input `char` is a valid punctuation character in Rust.
    - `Ident::new` ensures that the input `str` is a valid identifier in Rust.
    - Lifetimes in proc macros are now represented as two joint tokens - `Punct('\'', Spacing::Joint)` and `Ident("lifetime_name_without_quote")` similarly to multi-character operators.
- Stabilized APIs: None yet.

A bit of motivation for renaming (although it was already stated in the review comments):
- With my compiler frontend glasses on `Ident` is the single most appropriate name for this thing, *especially* if we are doing input validation on construction. `TokenTree::Ident` effectively wraps `token::Ident` or `ast::Ident + is_raw`, its meaning is "identifier" and it's already named `ident` in declarative macros.
- Regarding `Punct`, the motivation is that `Op` is actively misleading. The thing doesn't mean an operator, it's neither a subset of operators (there is non-operator punctuation in the language), nor superset (operators can be multicharacter while this thing is always a single character). So I named it `Punct` (first proposed in [the original RFC](https://github.com/rust-lang/rfcs/pull/1566), then [by @SimonSapin](https://github.com/rust-lang/rust/issues/38356#issuecomment-276676526)) , together with input validation it's now a subset of ASCII punctuation character category (`u8::is_ascii_punctuation`).

### CODE REVIEW

**Comment by petrochenkov** [comment]
@dtolnay  @alexcrichton @eddyb @nrc
Could you go through the comments marked with "REVIEW" and tell what you think?

File: src/libproc_macro/lib.rs
Comment (eddyb) [inline src/libproc_macro/lib.rs]: Also, `Word` is now `Term`, right?

File: src/libproc_macro/lib.rs
Comment (eddyb) [inline src/libproc_macro/lib.rs]: Validation seems really nice, I think previously it's been deferred until conversion to the compiler-internal forms is needed.

File: src/libproc_macro/lib.rs
Comment (eddyb) [inline src/libproc_macro/lib.rs]: I can confirm that this would not cause issues with #49219.

File: src/libproc_macro/lib.rs
Comment (eddyb) [inline src/libproc_macro/lib.rs]: Can this become `unreachable!()` now? As only a finite set of characters can be used.

File: src/libproc_macro/lib.rs
Comment (eddyb) [inline src/libproc_macro/lib.rs]: If we do keep it `Copy`, that means we can *never* soundly offer an `&self -> &str` API. See https://github.com/rust-lang/rust/issues/38356#issuecomment-386786645.

**Comment by alexcrichton** [comment]
All the changes here look great to me, thanks @petrochenkov! While we've already FCP'd for stability in https://github.com/rust-lang/rust/issues/38356 I think the `Ident` name is probably better than `Term` and I'm personally ambivalent on `Op` or `Punct`, but am happy to go with `Punct`.

I'd prefer to land this all at once if possible to reduce churn in the ecosystem, so I'd be mostly in favor of going ahead and stabilizing everything in this PR (if you're ok with that). I don't mind doing the work myself for the APIs here and pushing those up (annotation-wise). Along those lines, want to remove `as_str` as well from this?

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: I think this doesn't match the original implementation, when we first started we were sure to return an error I thought...

In any case I'd think it's a bug that this panics, it should definitely be returning an error (but we can fix that later)

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: Yeah this is currently the main way to go from a `TokenTree` to a `TokenStream`, and I think we also accidentally stabilized it already so we're required to keep it.

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: Yeah I think we're likely to always want `Span` to be `Copy`

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: I suspect impls like these will grow and come to exist over time yeah, that's definitely the direction that libstd has taken.

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: I think this is less critical than `Span` to be copy, and wouldn't mind going either way here. Making this not-`Copy` is probably the most conservative thing to do

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: Yes let's stick to `char` for maximal future flexibility.

File: src/libproc_macro/lib.rs
Comment (alexcrichton) [inline src/libproc_macro/lib.rs]: Yes for future flexibility I think `char`'s the way to go here

**Review by dtolnay — DISMISSED** [review]
The reason we decided against TokenTree::Ident was because the matcher `$i:ident` does not accept lifetimes. I would be opposed to calling it Ident here if it can hold lifetimes and `$i:ident` cannot.

File: src/libproc_macro/lib.rs
Comment (dtolnay) [reply src/libproc_macro/lib.rs]: I personally don't mind if we remove the Copy here.

File: src/libproc_macro/lib.rs
Comment (dtolnay) [reply src/libproc_macro/lib.rs]: I agree, let's go with no Copy.

**Comment by petrochenkov** [comment]
@dtolnay 
I'd be totally okay with accepting lifetimes in `$i: ident` (with or without dropping `$l: lifetime`) if we are ready to cement lifetimes as a "lexical class inside identifiers" as opposed to "`'` plus an identifier" (which proc macro API already does).
There's internals thread about this https://internals.rust-lang.org/t/pre-rfc-splitting-lifetime-into-two-tokens/6716/10. I don't think the original suggestion is a good idea anymore, but see https://internals.rust-lang.org/t/pre-rfc-splitting-lifetime-into-two-tokens/6716/11 in particular about the "lexical class inside identifiers" alternative.

The key questions here are probably:
- Is it important to discern between identifiers and lifetimes in procedural macros? Why?
- Is it important to discern between identifiers and lifetimes in declarative macros? Why?

**Comment by petrochenkov** [comment]
@alexcrichton 
>I'd prefer to land this all at once if possible to reduce churn in the ecosystem, so I'd be mostly in favor of going ahead and stabilizing everything in this PR (if you're ok with that). I don't mind doing the work myself for the APIs here and pushing those up (annotation-wise). Along those lines, want to remove as_str as well from this?

Sure, I'll add `stable` annotations and remove `as_str`.

**Comment by eddyb** [comment]
> Is it important to discern between identifiers and lifetimes in declarative macros? Why?

Probably to parse anything looking like a list of generics, if you want to tell apart `'a` and `T`.

**Comment by dtolnay** [comment]
> It's possible to detect lifetimes in macros left sides if we keep `lifetime` as a subset of `ident`

I would need to play with this in nightly before I would feel comfortable committing to it. What I wouldn't want is:

- We stabilize TokenTree::Ident under the expectation that we will make `$:ident` accept lifetimes.
- We change `$:ident` in decl macros to accept lifetimes, while keeping `$:lifetime`.
- After some use we discover this is worse this way than having `$:ident` and `$:lifetime` disjoint.
- We revert `$:ident` to reject lifetimes.
- Now proc macro and decl macro are inconsistent in an obnoxious way.

**Comment by dtolnay** [comment]
One place I can imagine running into trouble is if a macro somehow needs to handle optional lifetimes on a reference.

```rust
macro_rules! m {
    (& $($p:lifetime)? $i:ident) => {}
}

m!(&foo);
m!(&'a foo);
```

```
error: local ambiguity: multiple parsing options: built-in NTs lifetime ('p') or ident ('i').
 --> src/main.rs:4:5
  |
4 | m!(&foo);
  |     ^^^
```

**Comment by petrochenkov** [comment]
@dtolnay 
>local ambiguity: multiple parsing options: built-in NTs lifetime ('p') or ident ('i').

This is a weird error. I never investigated how our macro matching works in detail, but I expected it to be modeled after regular expressions (greedy by default) and work in both cases, whether `lifetime` is a subset of `ident` or not.
```rust
extern crate regex;
use regex::Regex;

fn main() {
    let not_subset = Regex::new(r"a?b").unwrap();
    assert!(not_subset.is_match("ab"));
    assert!(not_subset.is_match("b"));
    let subset = Regex::new(r"a?\w").unwrap();
    assert!(subset.is_match("ab"));
    assert!(subset.is_match("aa"));
    assert!(subset.is_match("b"));
    assert!(subset.is_match("a"));
}
```
http://play.rust-lang.org/?gist=372d5a5f25219670dabc89a1dfb536ba&version=nightly&mode=debug

**Comment by eddyb** [comment]
I've nominated for lang team discussion of https://github.com/rust-lang/rust/pull/50473#issuecomment-387255189.

cc @DanielKeep who has some interest in parsing generic declarations in `macro_rules`.

**Comment by petrochenkov** [comment]
@dtolnay 
If `ident` and `lifetime` are kept separated in declarative macros, I'd prefer to keep them separated in procedural macros as well.
Probably in the same way like this PR already does with raw identifiers, i.e. through constructors, not types:
```rust
Term::ident("foo"); // Make an identifier `foo`
Term::raw_ident("foo"); // Make a raw identifier `r#foo`
Term::lifetime("foo"); // Make a lifetime `foo`, note the lack of starting `'` (*)
// Term::raw_lifetime("foo") // Maybe, some day
```
(*) A long time (but low priority) desire of mine was to drop `'` from lifetime identifiers internally in the compiler, it's not really needed for anything, except for pretty-printing.
Today I implemented this (https://github.com/petrochenkov/rust/commit/7f83489cb3b60eb062fbe919dfd6164f756145c6) and it looks pretty good (and also avoids two allocations per lifetime and reduces the number of interned string), so I'm quite sure we can avoid "string typing" for lifetimes in proc macro API and accept strings without starting `'` in `Term::lifetime` without extra conversions.

**Comment by dtolnay** [comment]
@petrochenkov that sounds fine to me.

**Comment by eddyb** [comment]
@petrochenkov FWIW I like the idea of a `'` token joined to an identifier, but I'm not sure if it's viable.

**Comment by durka** [comment]
See my comment https://github.com/rust-lang/rust/pull/50385#issuecomment-387257697 about the issue with making lifetimes match as `:ident` (TLDR it's breaking).

**Comment by alexcrichton** [comment]
If `Term` isn't being renamed to `Ident` (as it sounds like due to lifetimes) then is there still enough motivation to change `Op` to `Punct`?

**Comment by dtolnay** [comment]
Yes I prefer `Punct` over `Op` as motivated in the pull request description. We are converging on my original naming choices -- https://github.com/rust-lang/rust/pull/40939#issuecomment-303300697.

**Comment by joshtriplett** [comment]
@petrochenkov We discussed this a bit in the lang team meeting. Initial impressions: it makes sense to be able to match a lifetime by matching a `'` followed by an `$ident`, but `$ident` alone should *not* match the `'`, in either macros-by-example or proc macros.

We *certainly* shouldn't rush to shoehorn `'` into `$ident` just because we don't have a specific way to match lifetimes instead.

If a macro wants to emit a lifetime, I don't see a problem with emitting `'` followed by `$ident`, but that doesn't seem like a reason to let the *language* accept comments or whitespace in between.

We'd still like to hear from @eddyb, who wasn't in the lang team meeting today.

**Comment by cramertj** [comment]
I'm in favor of the original change @petrochenkov proposed in the [splitting lifetimes into two tokens Pre-RFC thread](https://internals.rust-lang.org/t/pre-rfc-splitting-lifetime-into-two-tokens/6716). I have personally wanted to be able to construct a lifetime from `'$my_ident` in macros, and it makes sense to me to think of `'foo` as a lifetime symbol followed by an ident.

**Comment by petrochenkov** [comment]
@joshtriplett @cramertj 
Interesting! This sympathy to splitting `'ident` into two tokens was somewhat unexpected to me.

I see two ways forward from this position:
1. Stabilize `lifetime` in declarative macros (https://github.com/rust-lang/rust/pull/50385) and `Term::lifetime` in procedural macros right now, without preceding split of `'ident` into two tokens.
I think we'll be able to perform this split later in backward compatible way - `lifetime` in declarative macros will match two tokens and `Term::lifetime` in procedural macros will generate two tokens for the parser like `Literal::i32_unsuffixed(arg: i32)` and friends will generate two tokens (`-` and numeric literal) for negative `arg`s.
2. Split `'ident` into two tokens right now, before stabilizing `lifetime` and proc macro API 1.2 (without introducing observable changes into non-proc-macro parts of the language).

For `1.` we may want to also tweak how `Spacing` works, e.g. generate `Joint` spacing even if the following token is not a punctuation character (@dtolnay, this shouldn't create inconveniences for tokenstream-based parsers like `syn`, right?).
Then `'ident` will be somewhat similar to `+=` (`[Punct('\'', Joint), Term("ident")]` and `[Punct('+', Joint), Punct('=', Alone)]`) and will "combine" only with correct spacing.

**Comment by petrochenkov** [comment]
Ugh.
`'lifetime` currently matches `tt` in declarative macros because it's a single token (stable, since 1.0).
`'lifetime` as two tokens won't match `tt`.

**Comment by cramertj** [comment]
> 'lifetime as two tokens won't match tt.

Gah! I don't suppose it's practical to special-case this, is it?

**Comment by petrochenkov** [comment]
@cramertj 
>I don't suppose it's practical to special-case this, is it?

I'm not sure how many hacks it will require exactly.
We'll need to at least redefine `tt` as "token tree or `'` + `IDENT`" and actually construct a tree with empty delimiter from those two tokens (equivalent to `Group(Delimiter::None, [Punct('\'', Joint), Term("foo")])`).

**Comment by petrochenkov** [comment]
@eddyb
>Doesn't (+, Joint) followed by (=, Alone) also create a single += tt?

That's a good observation.
Perhaps imitating combined operators is the way to go for lifetimes as well?
Proc macros see `'ident` as two joint tokens (like `+=`), the rest of the language sees `'ident` as a single token (like `+=`) at least for now.
(That's a combination of `1.` and `2.` from https://github.com/rust-lang/rust/pull/50473#issuecomment-388200443, basically.)

The logic is the same as with operators - we break tokens into "basic components" for proc macros to provide compatibility with any possible future extensions (+ also give possibility to create custom operators).

**Comment by eddyb** [comment]
@dtolnay I remember discussing (at the Berlin all-hands) the similar problem of floating-point literals, and potentially decomposing them into smaller tokens, and you mentioning the need for a correspondence between declarative and procedural macros' token models.
Does https://github.com/rust-lang/rust/pull/50473#issuecomment-388336736 invalidate that correspondence? Sadly I don't recall the exact details.

**Comment by petrochenkov** [comment]
PR is updated, comments are addressed, PR description is updated as well.

The last commit breaks up lifetimes into two tokens (`Punct('\'', Spacing::Joint)` and `Ident("lifetime_name_without_quote")`) similarly to how compound operators are broken up.
This should be future-compatible with any lifetime-related changes in declarative macros and "proper language" in general (`'$ident`, `lifetime a`, etc).
See https://github.com/rust-lang/rust/pull/50473#issuecomment-388160288 and below for relevant discussion.

File: src/libproc_macro/lib.rs
Comment (petrochenkov) [inline src/libproc_macro/lib.rs]: At first I started reporting `Joint` for any non-whitespace token immediately following a punctuation character (for token streams coming from lexer/libsyntax), but it felt wrong enough to revert the change.
It's quite strange to give proc macros ability to discern between e.g. `+x` and `+ x`, something that the "proper language" strives to never do, and combining anything with e.g. doc comments (`+/// Doc`) is plain insane.
So I made a very targeted tweak instead and now report non-punctuation `Joint` only for `'`s followed by identifiers.

Since jointness is more "revealed" than "reported", I believe we'll be able to backward-compatibly "reveal more" in the future if it becomes necessary.

**Comment by alexcrichton** [comment]
Awesome, thanks so much for your continued work on this @petrochenkov! Everything here looks great to me.

Given the discovery of https://github.com/rust-lang/rust/issues/50725 I think we may want to hold off on the stabilization here? I'm hoping we can either stabilize all-in-one-go or perhaps start with the module system names first.

We've got a long time before beta branches, though. Do you think we can track down that issue before the next beta release? If so I think it's fine to go ahead an stabilize

**Comment by petrochenkov** [comment]
>Do you think we can track down that issue before the next beta release?

Fix for https://github.com/rust-lang/rust/issues/50725 is on the top of my "queue of things to implement" right now, I discovered it while trying to prepare a stabilization PR for `use_extern_macros`.

>Given the discovery of #50725 I think we may want to hold off on the stabilization here?

`proc_macro` as a language feature is indeed dependent on `use_extern_macros`, but this API isn't, it can be equally well used for already stable proc macro derives. So, in principle it can be stabilized even without `proc_macro`.

That said, there's no reason to hurry with stabilization, so I'd rather remove the stabilization commit and land this PR sooner to give people on nightly change to use the updated version and also unblock https://github.com/rust-lang/rust/pull/49219.

**Comment by alexcrichton** [comment]
Sounds good to me! r=me with the stabilization commit backed out

**Comment by kennytm** [comment]
@bors p=1

Unblocks #49219

**Pull request merged**
